pytest_operator/plugin.py

```diff
@@ -342,46 +343,134 @@ def download_resource(self, resource, destination: Path):
 
-    _instance = None  # store instance so we can tell if it's been used yet
+    _instance = None  # store instance, so we can tell if it's been used yet
 
     def __init__(self, request, tmp_path_factory):
         self.request = request
-        self.tmp_path: Path = tmp_path_factory.mktemp(self.default_model_name)
-        log.info(f"Using tmp_path: {self.tmp_path}")
+        self._tmp_path_factory = tmp_path_factory
+        self._global_tmp_path = None
 
         # Flag indicating whether all subsequent tests should be aborted.
         self.aborted = False
 
         # Flag for using destructive mode or not for charm builds.
         self.destructive_mode = request.config.option.destructive_mode
 
+        # Config options to determine first model specs used by tests
+        self._init_cloud_name = request.config.option.cloud
+        self._init_model_name = request.config.option.model
+        self._init_keep_model = request.config.option.keep_models
+
         # These may be modified by _setup_model
-        self.cloud_name = request.config.option.cloud
         self.controller_name = request.config.option.controller
-        self.model_name = request.config.option.model
-        self.keep_model = request.config.option.keep_models
         self.model_config = request.config.option.model_config
 
         # Flag for enabling the juju-crashdump
         self.crash_dump = not request.config.option.no_crash_dump
         self.crash_dump_output = request.config.option.crash_dump_output
 
         # These will be set by _setup_model
-        self.model_full_name = None
-        self.model = None
         self.jujudata = None
-        self._controller = None
+        self._controller: Optional[Controller] = None
+
+        # maintains a set of all models connected by this fixture
+        self.current_model: Optional[ModelKey] = None
+        self._models: MutableMapping[ModelKey, ModelState] = {}
+
+    def switch(self, model_key: ModelKey) -> Model:
+        """
+        Analog to `juju switch` where the focus of the current model is moved.
+        """
+        if model_key in self.models:
+            self.current_model = model_key
+        else:
+            raise ModelNotFoundError(f"{model_key} not found")
+
+        return self.model
+
+    @property
+    def tmp_path(self) -> Path:
+        tmp_path = self._global_tmp_path
+        current_state = self._models.get(self.current_model)
+        if current_state and current_state.tmp_path is None:
+            tmp_path = self._tmp_path_factory.mktemp(self.current_model.model)
+            current_state.tmp_path = tmp_path
+        elif current_state and current_state.tmp_path:
+            tmp_path = current_state.tmp_path
+        elif not tmp_path:
+            tmp_path = self._global_tmp_path = self._tmp_path_factory.mktemp(
+                self.default_model_name
+            )
+
+        log.info(f"Using tmp_path: {tmp_path}")
+        return tmp_path
```

Comment:
create and store the `tmp_path` within the `self._models` ModelState based on the `current_model` pointer